src/attr/_make.py

```diff
@@ -1873,64 +1874,132 @@ def _add_eq(cls, attrs=None):
+        lines.append(
+            "    return f'{}({})'".format(cls_name_fragment, repr_fragment)
+        )
+        lines.append("  finally:")
+        lines.append("    working_set.remove(id(self))")
+        script = "\n".join(lines)
+        return _make_method("__repr__", script, unique_filename, globs=globs)
 
-    # Figure out which attributes to include, and which function to use to
-    # format them. The a.repr value can be either bool or a custom callable.
-    attr_names_with_reprs = tuple(
-        (a.name, repr if a.repr is True else a.repr)
-        for a in attrs
-        if a.repr is not False
-    )
 
-    def __repr__(self):
+else:
+
+    def _make_repr(attrs, ns, _):
         """
-        Automatically created by attrs.
+        Make a repr method that includes relevant *attrs*, adding *ns* to the
+        full name.
         """
-        try:
-            working_set = _already_repring.working_set
-        except AttributeError:
-            working_set = set()
-            _already_repring.working_set = working_set
 
-        if id(self) in working_set:
-            return "..."
-        real_cls = self.__class__
-        if ns is None:
-            qualname = getattr(real_cls, "__qualname__", None)
-            if qualname is not None:
-                class_name = qualname.rsplit(">.", 1)[-1]
-            else:
-                class_name = real_cls.__name__
-        else:
-            class_name = ns + "." + real_cls.__name__
+        # Figure out which attributes to include, and which function to use to
+        # format them. The a.repr value can be either bool or a custom
+        # callable.
+        attr_names_with_reprs = tuple(
+            (a.name, repr if a.repr is True else a.repr)
+            for a in attrs
+            if a.repr is not False
+        )
 
-        # Since 'self' remains on the stack (i.e.: strongly referenced) for the
-        # duration of this call, it's safe to depend on id(...) stability, and
-        # not need to track the instance and therefore worry about properties
-        # like weakref- or hash-ability.
-        working_set.add(id(self))
-        try:
-            result = [class_name, "("]
-            first = True
-            for name, attr_repr in attr_names_with_reprs:
-                if first:
-                    first = False
+        def __repr__(self):
+            """
+            Automatically created by attrs.
+            """
+            try:
+                working_set = _already_repring.working_set
+            except AttributeError:
+                working_set = set()
+                _already_repring.working_set = working_set
+
+            if id(self) in working_set:
+                return "..."
+            real_cls = self.__class__
+            if ns is None:
+                qualname = getattr(real_cls, "__qualname__", None)
+                if qualname is not None:
+                    class_name = qualname.rsplit(">.", 1)[-1]
```

Comment:
```suggestion
                if qualname is not None:  # pragma: no cover
                    # This case only happens on Python 3.5 and 3.6. We exclude
                    # it from coverage, because we don't want to slow our test
                    # suite by running them under coverage too for this one
                    # line.
                    class_name = qualname.rsplit(">.", 1)[-1]
```